File: Python/src/graph_basic.py
# BFS, DFS, Union-Find (disjoint set)
from collections import deque
from typing import Dict, List, Set

def bfs(graph: Dict[int, List[int]], start: int) -> List[int]:
    q = deque([start])
    visited = set([start])
    order = []
    while q:
        u = q.popleft()
        order.append(u)
        for v in graph.get(u, []):
            if v not in visited:
                visited.add(v)
                q.append(v)
    return order

def dfs_iter(graph: Dict[int, List[int]], start: int) -> List[int]:
    stack = [start]
    visited = set()
    order = []
    while stack:
        u = stack.pop()
        if u in visited: 
            continue
        visited.add(u)
        order.append(u)
        for v in reversed(graph.get(u, [])):
            if v not in visited:
                stack.append(v)
    return order

class UnionFind:
    def __init__(self, n):
        self.parent = list(range(n))
        self.rank = [0]*n
    def find(self, x):
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]
    def union(self, x, y):
        rx, ry = self.find(x), self.find(y)
        if rx == ry: return False
        if self.rank[rx] < self.rank[ry]:
            self.parent[rx] = ry
        elif self.rank[rx] > self.rank[ry]:
            self.parent[ry] = rx
        else:
            self.parent[ry] = rx
            self.rank[rx] += 1
        return True

File: Java/src/GraphBasic.java
import java.util.*;

public class GraphBasic {

    public static List<Integer> bfs(Map<Integer, List<Integer>> g, int start) {
        Queue<Integer> q = new ArrayDeque<>();
        Set<Integer> seen = new HashSet<>();
        List<Integer> order = new ArrayList<>();
        q.add(start); seen.add(start);

        while (!q.isEmpty()) {
            int u = q.poll();
            order.add(u);
            for (int v : g.getOrDefault(u, Collections.emptyList())) {
                if (!seen.contains(v)) {
                    seen.add(v);
                    q.add(v);
                }
            }
        }
        return order;
    }

    public static List<Integer> dfsIter(Map<Integer, List<Integer>> g, int start) {
        Stack<Integer> st = new Stack<>();
        Set<Integer> seen = new HashSet<>();
        List<Integer> order = new ArrayList<>();
        st.push(start);

        while (!st.empty()) {
            int u = st.pop();
            if (seen.contains(u)) continue;
            seen.add(u);
            order.add(u);
            List<Integer> neigh = g.getOrDefault(u, Collections.emptyList());
            for (int i = neigh.size()-1; i >= 0; i--)
                st.push(neigh.get(i));
        }
        return order;
    }
}
